iree/turbine/kernel/wave/index_sequence_analysis.py

```diff
@@ -462,6 +379,226 @@ def set_node_index(
+):
+    """
+    Specialize the index sequence with the given substitutions.
+    """
+    return {dim: seq.subs(subs) for dim, seq in index.items()}
+
+
+def populate_mma_sources(
+    node: MMA,
+    mma_index: dict[MMA, dict[IndexSymbol, int]],
+    hardware_constraint: HardwareConstraint,
+):
+    """
+    Initialize the sources with the LHS, RHS, ACC and MMA node
+    and their index sequences and vector shapes. These will
+    be propagated to the rest of the graph.
+    """
+    index: dict[IndexSymbol, IndexSequence] = {}
+    mapping = mma_index[node]
+    for dim, dim_index in mapping.items():
+        index[dim] = hardware_constraint.apply(
+            dim, dim_index, None, None, True, node.mma_type
+        )
+    node.index = combine_indices(node.index, index)
+    return [
+        (
+            get_custom(node.lhs),
+            specialize_index(index, {MMA_LHS: 1, MMA_RHS: 0, MMA_ACC: 0}),
+            node.vector_shapes,
+        ),
+        (
+            get_custom(node.rhs),
+            specialize_index(index, {MMA_LHS: 0, MMA_RHS: 1, MMA_ACC: 0}),
+            node.vector_shapes,
+        ),
+        (
+            get_custom(node.acc),
+            specialize_index(index, {MMA_LHS: 0, MMA_RHS: 0, MMA_ACC: 1}),
+            node.vector_shapes,
+        ),
+        (
+            node,
+            specialize_index(index, {MMA_LHS: 0, MMA_RHS: 0, MMA_ACC: 1}),
+            node.vector_shapes,
+        ),
+    ]
+
+
+def populate_non_mma_sources(
+    node: Read | Write,
+    hardware_constraint: HardwareConstraint,
+    workgroup_constraints: list[WorkgroupConstraint],
+):
+    """
+    Initialize the sources with the read and/or write nodes
+    and their index sequences and vector shapes. These will
+    be propagated to the rest of the graph.
+    """
+    index: dict[IndexSymbol, IndexSequence] = {}
+    for dim in node.indexing_dims:
+        elements_per_thread = (
+            1
+            if not is_contiguous_dim(
+                dim, node.indexing_dims, hardware_constraint.vector_shapes
+            )
+            else node.elements_per_thread
+        )
+        stride = compute_stride(
+            node.indexing_dims, hardware_constraint.vector_shapes, dim
+        )
+        wg_constraint = [x for x in workgroup_constraints if x.dim == dim]
+        if not wg_constraint:
+            continue
+        index[dim] = hardware_constraint.apply(
+            dim,
+            wg_constraint[0].workgroup_dim,
+            elements_per_thread,
+            stride,
+            False,
+            None,
+        )
+    return [(node, index, hardware_constraint.vector_shapes)]
+
+
+def combine_indices(
+    thread_independent_index: dict[IndexSymbol, IndexSequence],
+    thread_dependent_index: dict[IndexSymbol, IndexSequence],
+) -> dict[IndexSymbol, IndexSequence]:
+    combined_index = {k: v for k, v in thread_independent_index.items()}
+    for k in combined_index:
```

Comment:
Would be nice to add some docs, to different examples/cases on when we'd add the start offsets. (perhaps read -> read or offseted_read + mma?)

Reply:
In this PR, we split the index assignment into 2 phases. The first is a thread-independent index assignment where we set the indices based on work group constraints and tiling constraints and anything that in general does not have any thread level dependence. Once we have this index then we go through and propagate the thread dependent index which comes either from MMA nodes or if there are no MMA nodes it comes from reads and writes and we then add the thread dependent index to the thread independent index. So we are always going to be adding these two offsets together.